src/sema/diagnostics.rs

```diff
@@ -2,51 +2,117 @@ use super::ast::{Diagnostic, Level, Namespace};
+impl Diagnostics {
+    pub fn any_errors(&self) -> bool {
+        self.has_error
+    }
 
-    for msg in &ns.diagnostics {
-        if msg.level == Level::Debug && !debug {
-            continue;
+    pub fn len(&self) -> usize {
+        self.contents.len()
+    }
+
+    pub fn iter(&self) -> Iter<Diagnostic> {
+        self.contents.iter()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.contents.is_empty()
+    }
+
+    pub fn push(&mut self, diagnostic: Diagnostic) {
+        if matches!(diagnostic.level, Level::Error) {
+            self.has_error = true;
         }
+        self.contents.push(diagnostic);
+    }
 
-        let diagnostic = convert_diagnostic(msg, &file_id);
+    pub fn extend(&mut self, diagnostics: Vec<Diagnostic>) {
+        if !self.has_error {
+            self.has_error = diagnostics.iter().any(|m| m.level == Level::Error);
+        }
+        self.contents.extend(diagnostics);
+    }
 
-        term::emit(&mut writer.lock(), &config, &files, &diagnostic).unwrap();
+    pub fn append(&mut self, diagnostics: &mut Vec<Diagnostic>) {
+        if !self.has_error {
+            self.has_error = diagnostics.iter().any(|m| m.level == Level::Error);
+        }
+        self.contents.append(diagnostics);
     }
-}
 
-/// Print the diagnostics to stdout with plain formatting
-pub fn print_diagnostics_plain(cache: &FileResolver, ns: &Namespace, debug: bool) {
-    let (files, file_id) = convert_files(ns, cache);
+    pub fn first_error(&self) -> String {
+        match self.contents.iter().find(|m| m.level == Level::Error) {
+            Some(m) => m.message.to_owned(),
+            None => panic!("no errors found"),
+        }
+    }
 
-    let config = term::Config::default();
+    pub fn count_warnings(&self) -> usize {
+        self.contents
+            .iter()
+            .filter(|&x| x.level == Level::Warning)
+            .count()
+    }
 
-    for msg in &ns.diagnostics {
-        if msg.level == Level::Debug && !debug {
-            continue;
+    pub fn first_warning(&self) -> &Diagnostic {
+        self.contents
+            .iter()
+            .find_or_first(|&x| x.level == Level::Warning)
+            .unwrap()
+    }
+
+    pub fn warnings(&self) -> Vec<&Diagnostic> {
+        let mut res = Vec::new();
+        for elem in &self.contents {
+            if elem.level == Level::Warning {
+                res.push(elem);
+            }
         }
 
-        let diagnostic = convert_diagnostic(msg, &file_id);
+        res
+    }
 
-        let mut buffer = RawBuffer::new();
+    pub fn errors(&self) -> Vec<&Diagnostic> {
+        let mut vec = Vec::new();
+        for diag in &self.contents {
+            if matches!(diag.level, Level::Error) {
+                vec.push(diag);
+            }
+        }
+        vec
+    }
 
-        term::emit(&mut buffer, &config, &files, &diagnostic).unwrap();
+    pub fn assert_message_in_warnings(&self, message: &str) -> bool {
```

Comment:
This function does not assert, it should be called `fn contains_message_in_warnings()` or `fn warning_contains()`